qutip/tests/test_qubitcircuit.py

```diff
@@ -203,31 +203,88 @@ def test_add_state(self):
         qc.add_state("-", targets=[1])
 
         assert_(qc.input_states[0] == "0")
-        assert_(qc.input_states[2] == None)
+        assert_(qc.input_states[2] is None)
         assert_(qc.output_states[1] == "+")
 
         qc1 = QubitCircuit(10)
 
         qc1.add_state("0", targets=[2, 3, 5, 6])
-        qc1.add_state("+", targets=[1,4,9])
-        qc1.add_state("A", targets=[1,4,9], state_type="output")
-        qc1.add_state("A", targets=[1,4,9], state_type="output")
+        qc1.add_state("+", targets=[1, 4, 9])
+        qc1.add_state("A", targets=[1, 4, 9], state_type="output")
+        qc1.add_state("A", targets=[1, 4, 9], state_type="output")
         qc1.add_state("beta", targets=[0], state_type="output")
-        assert_(qc1.input_states[0] == None)
+        assert_(qc1.input_states[0] is None)
 
         assert_(qc1.input_states[2] == "0")
         assert_(qc1.input_states[3] == "0")
         assert_(qc1.input_states[6] == "0")
         assert_(qc1.input_states[1] == "+")
         assert_(qc1.input_states[4] == "+")
 
-        assert_(qc1.output_states[2] == None)
+        assert_(qc1.output_states[2] is None)
         assert_(qc1.output_states[1] == "A")
         assert_(qc1.output_states[4] == "A")
         assert_(qc1.output_states[9] == "A")
 
         assert_(qc1.output_states[0] == "beta")
 
+    def test_exceptions(self):
+        """
+        Text exceptions are thrown correctly for inadequate inputs
+        """
+        qc = QubitCircuit(2)
+        for gate in ["X", "Y", "Z", "S", "T"]:
+            assert_raises(ValueError, qc.add_gate, gate,
+                          targets=[1], controls=[0])
```

Comment:
In new `pytest`-style, the `for` loop would be done with parametrised tests and the raising test uses the `pytest.raises` context manager, so you get easier-to-read code (and it's selectable if you only want to test a certain gate) like
```python
@pytest.mark.parametrize('gate', ['X', 'Y', 'Z', 'S' ,'T'])
def test_exceptions(self, gate):
    qc = QubitCircuit(2)
    with pytest.raises(ValueError):
        qc.add_gate(gate, targets=[1], controls=[0])
```
Again, thinking of the test log, this way is nice because the test log will tell you exactly which gate failed - `pytest` adds the parametrised arguments into the test name.